### PR TITLE
Feature/RegistryApi

### PR DESCRIPTION
**Pull request by KartaviK**

### CODE REVIEW

File: src/Reestr/Rep/Response.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Class Response
+ *
+ * @package Wearesho\Bobra\Ubki\Reestr\Rep
+ */
+class Response extends Reestr\Response implements ResponseInterface
+{
+    use ResponseTrait;
+
+    const ERROR_CRITICAL = 'CRITICAL';
```
Comment (Horat1us) [inline src/Reestr/Rep/Response.php]: missing visibility

File: src/Reestr/Rep/Request.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+namespace Wearesho\Bobra\Ubki\Reestr\Rep;
+
+use Wearesho\Bobra\Ubki\Reestr;
+
+/**
+ * Class Request
+ *
+ * @package Wearesho\Bobra\Ubki\Reestr\Rep
+ */
+class Request extends Reestr\Request
```
Comment (Horat1us) [inline src/Reestr/Rep/Request.php]: What purpose of this class?

File: src/Reestr/Rep/ResponseInterface.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public function getInn(): int;
+
+    /**
+     * Note text
+     *
+     * @return string
+     */
+    public function getRemark(): string;
+
+    public function isErrorCritical(): bool;
```
Comment (Horat1us) [inline src/Reestr/Rep/ResponseInterface.php]: Is Error always exists?

File: src/Reestr/Rep/ResponseInterface.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * @package Wearesho\Bobra\Ubki\Reestr\Rep
+ */
+interface ResponseInterface extends Reestr\ResponseInterface
+{
+    /**
+     * Reestr type
+     *
+     * @return string
+     */
+    public function getErtype(): string;
```
Comment (Horat1us) [inline src/Reestr/Rep/ResponseInterface.php]: Is error always exists?

File: src/Reestr/Request.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    /**
+     * Validate type of request
+     *
+     * @param string $todo
+     *
+     * @return bool
+     */
+    protected function validateTodo(string $todo): bool
```
Comment (Horat1us) [inline src/Reestr/Request.php]: validators should have return type void and throw some exceptions

File: src/Reestr/RequestInterface.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace Wearesho\Bobra\Ubki\Reestr;
+
+
```
Comment (Horat1us) [inline src/Reestr/RequestInterface.php]: Missing phpDoc

File: src/Reestr/RequestTrait.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    /**
+     * @inheritdoc
+     */
+    public function getIdalien(): string
+    {
+        return $this->idalien;
+    }
+
+    protected function validateIndate(\DateTimeInterface $indate): bool
```
Comment (Horat1us) [inline src/Reestr/RequestTrait.php]: validators should have void return type

File: src/Reestr/ResponseInterface.php [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    /**
+     * Authorization session key
+     *
+     * @return string
+     */
+    public function getSessid(): string;
+
+    /**
+     * State of sent report
+     * r - обработан, n - передан, i - создан, b - забракован, e - ошибка записи в базу
```
Comment (Horat1us) [inline src/Reestr/ResponseInterface.php]: Description have to be moved to constants

File: src/Reestr/ResponseInterface.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * Subcomponent
+     * (IDENT, URIDENT, LINKED, WORK, DOC, ADDR, FOTO, CRDEAL, SUSD, CONT...)
+     *
+     * @return string
+     */
+    public function getItem(): string;
+
+    public function isStateProcessed(): bool;
```
Comment (Horat1us) [inline src/Reestr/ResponseInterface.php]: This method is redundant

File: src/Reestr/Service.php [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+    /** @var Ubki\Authorization\ProviderInterface */
+    protected $authProvider;
+
+    /** @var Log\LoggerInterface */
+    protected $logger;
+
+    public function __construct(
+        ConfigInterface $config,
+        GuzzleHttp\ClientInterface $client,
```
Comment (Horat1us) [inline src/Reestr/Service.php]: Authorization provider have to be before GuzzleClient

File: src/Reestr/Service.php [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        /** @var GuzzleHttp\Psr7\Response $httpResponse */
+        $httpResponse = $this->client->send($guzzleRequest);
+        $responseBody = $httpResponse->getBody()->__toString();
+
+        if (!$this->config->isTestMode()) {
+            $fileUrl = $responseBody;
+            $responseBody = $this->getFile($fileUrl)->getBody()->__toString();
+        }
+
+        $xml = simplexml_load_string($responseBody);
```
Comment (Horat1us) [inline src/Reestr/Service.php]: what will be if xml is not valid? (HTML etc.)

**Review by Horat1us — CHANGES_REQUESTED** [review]

File: src/Reestr/Rep/ResponseInterface.php (hunk 4 above)
Comment (KartaviK) [reply src/Reestr/Rep/ResponseInterface.php]: This method return type of reestr. For error type have method getCrytical(). And this attr always exist.

from specification:
image

File: src/Reestr/Rep/ResponseInterface.php (hunk 3 above)
Comment (KartaviK) [reply src/Reestr/Rep/ResponseInterface.php]: Attribute with error type always exist in response, but can be empty. I add function isNoError() for empty case in later commit

File: src/Push/Registry/Rep/Request.php [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+{
+    public function __construct(
+        \DateTimeInterface $indate,
+        string $idout = "",
+        string $idalien = ""
+    ) {
+        parent::__construct(
+            Registry\Type::REP,
+            $indate,
+            $idout,
```
Comment (Horat1us) [inline src/Push/Registry/Rep/Request.php]: camelCase have to be used

File: src/Push/Registry/Rep/Response.php [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+class Response extends Registry\Response implements ResponseInterface
+{
+    use ResponseTrait;
+
+    public function __construct(
+        string $todo,
+        \DateTimeInterface $indate,
+        string $idout,
+        string $idalien,
+        string $sessid,
```
Comment (Horat1us) [inline src/Push/Registry/Rep/Response.php]: use camelCase and fullText names, without acronyms

File: src/Push/Registry/Service.php [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+                    (string)$context->item,
+                    (string)$context->ertype,
+                    (string)$context->crytical,
+                    (int)$context->inn,
+                    (string)$context->remark
+                );
+            case Type::BIL:
+                // TODO: need implement Bil request
+                break;
+        }
```
Comment (Horat1us) [inline src/Push/Registry/Service.php]: exception have to be thrown if unknown requestType received

File: src/Push/Registry/Service.php [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+    protected function catchException(GuzzleHttp\Exception\RequestException $exception): void
+    {
+        $xml = simplexml_load_string($exception->getResponse()->getBody()->__toString());
+
+        if ($xml === false || !isset($xml->prot)) {
+            // Not XML or invalid format XML
+            throw $exception;
+        }
+
+        throw new \Exception(
```
Comment (Horat1us) [inline src/Push/Registry/Service.php]: Use custom exception instead

File: src/Push/Registry/Type.php [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+namespace Wearesho\Bobra\Ubki\Push\Registry;
+
+/**
+ * Interface Type
+ *
+ * @package Wearesho\Bobra\Ubki\Push\Registry
+ */
+interface Type
+{
+    public const BIL = 'BIL';
```
Comment (Horat1us) [inline src/Push/Registry/Type.php]: Comments with description should be before const

**Review by Horat1us — CHANGES_REQUESTED** [review]

File: src/Push/Config.php [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        return $this->mode === static::MODE_PRODUCTION;
+    }
+
+    protected function validateMode(int $mode): void
+    {
+        $isInvalid =
+            $mode !== static::MODE_PRODUCTION &&
+            $mode !== static::MODE_TEST;
+
```
Comment (sashabeton) [inline src/Push/Config.php]: Use better `$isValid`

File: src/Push/ConfigInterface.php [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+ * Interface ConfigInterface
+ *
+ * @package Wearesho\Bobra\Ubki\Push
+ */
+interface ConfigInterface extends Authorization\ConfigInterface
+{
+    public const PRODUCTION_REESTR_URL = 'https://secure.ubki.ua/upload/in/reestrs.php';
+    public const PRODUCTION_PUSH_URL = '';
+
+    public const TEST_REESTR_URL = 'https://secure.ubki.ua:4040/upload/in/reestrs.php';
```
Comment (Horat1us) [inline src/Push/ConfigInterface.php]: registry_url should be used

**Review by Horat1us — CHANGES_REQUESTED** [review]

File: src/Push/Registry/RequestException.php [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+    protected $request;
+
+    public function __construct(
+        string $request,
+        int $code = 0,
+        \Throwable $previous = null
+    ) {
+        $this->request = $request;
+
+        parent::__construct('Reuest errors: ' . $request, $code, $previous);
```
Comment (sashabeton) [inline src/Push/Registry/RequestException.php]: Typo error in `Request`

File: src/Push/Registry/ResponseCollection.php [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Class ResponseCollection
+ *
+ * @package Wearesho\Bobra\Ubki\Push\Registry
+ */
+class ResponseCollection extends \ArrayObject implements \JsonSerializable
+{
+    public function __construct($input = array(), int $flags = 0, string $iteratorClass = "ArrayIterator")
+    {
```
Comment (sashabeton) [inline src/Push/Registry/ResponseCollection.php]: []

File: src/Push/Registry/Service.php [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+        /** @var GuzzleHttp\Psr7\Response $response */
+        $response = $this->client->send(
+            new GuzzleHttp\Psr7\Request(
+                $method = 'get',
+                $url,
+                []
+            )
+        );
+
+        return $response;
```
Comment (sashabeton) [inline src/Push/Registry/Service.php]: `return $response->getBody()->__toString()`
And change return type to string

File: src/Push/Registry/Service.php [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    /**
+     * @param string $url
+     *
+     * @return GuzzleHttp\Psr7\Response
+     * @throws GuzzleHttp\Exception\GuzzleException
+     */
+    private function getFile(string $url): GuzzleHttp\Psr7\Response
+    {
```
Comment (sashabeton) [inline src/Push/Registry/Service.php]: rename to `getFileContent`

File: src/Push/Registry/Service.php [hunk 23]
```diff
@@ -0,0 +1,257 @@
...
+        $urlFileResponse = $this->client->send($guzzleRequest);
+        $urlFile = $urlFileResponse->getBody()->__toString();
+
+        $this->validateUrl($urlFile);
+
+        try {
+            $responseBody = $this->getFile($urlFile)->getBody()->__toString();
+        } catch (GuzzleHttp\Exception\RequestException $exception) {
+            $this->catchException($exception);
+        }
```
Comment (sashabeton) [inline src/Push/Registry/Service.php]: Move logic of `catchException` method to `getFileContent`. And move try/catch

File: src/Push/Registry/Service.php [hunk 24]
```diff
@@ -0,0 +1,257 @@
...
+                );
+        }
+    }
+
+    /**
+     * @param GuzzleHttp\Exception\RequestException $exception
+     *
+     * @throws \Exception
+     */
+    protected function catchException(GuzzleHttp\Exception\RequestException $exception): void
```
Comment (sashabeton) [inline src/Push/Registry/Service.php]: Move logic to `getFileContent`

**Review by sashabeton — DISMISSED** [review]

File: src/Push/Config.php [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+        if ($isInvalid) {
+            // todo: remove unsupported-mode-exception to base directory
```
Comment (Horat1us) [inline src/Push/Config.php]: Isn't it in base directory already?

File: tests/Push/Registry/ResponseCollectionTest.php [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+    protected $responses;
+
+    /** @var Registry\ResponseCollection */
+    protected $testResponseCollection;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->responses = array(
```
Comment (Horat1us) [inline tests/Push/Registry/ResponseCollectionTest.php]: Array function should not be used

File: src/Push/EnvironmentConfig.php [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * Class EnvironmentConfig
+ *
+ * @package Wearesho\Bobra\Ubki\Push
+ */
+class EnvironmentConfig extends Environment\Config implements ConfigInterface
+{
+    use Authorization\EnvironmentConfigTrait;
+
```
Comment (Horat1us) [inline src/Push/EnvironmentConfig.php:17]: Prefix have to be specified by default

**Review by Horat1us — CHANGES_REQUESTED** [review]

**Review by Horat1us — APPROVED** [review]

File: src/Push/Registry/Response/State.php [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+    public const BLOCKED = 'b';
+    public const SQL_ERROR = 'e';
+
+    /**
+     * @param string $state
+     *
+     * @return State
+     * @throws \InvalidArgumentException
+     */
+    public static function instanceByValue(string $state): State
```
Comment (Horat1us) [inline src/Push/Registry/Response/State.php]: `createFromValue` should be used as name

Comment (Horat1us) [reply src/Push/Registry/Response/State.php]: Isn't it duplicates __construct from parent class?

File: src/Push/Registry/ResponseCollection.php [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+use Carbon\Carbon;
+
+use Wearesho\Bobra\Ubki\Push\Registry;
+
+/**
+ * Class ResponseCollection
+ *
+ * @package Wearesho\Bobra\Ubki\Push\Registry
+ */
+class ResponseCollection extends \ArrayObject implements \JsonSerializable
```
Comment (Horat1us) [inline src/Push/Registry/ResponseCollection.php]: Collection have responsibility of storing items.
Creating items isn't purpose of collections

File: src/Push/Registry/Service.php [hunk 30]
```diff
@@ -0,0 +1,194 @@
...
+        try {
+            $response = $this->client->send(
+                new GuzzleHttp\Psr7\Request(
+                    $method = 'get',
+                    $url,
+                    []
+                )
+            );
+        } catch (GuzzleHttp\Exception\RequestException $exception) {
+            $xml = simplexml_load_string($exception->getResponse()->getBody()->__toString());
```
Comment (Horat1us) [inline src/Push/Registry/Service.php]: getResponse() can return null

**Review by Horat1us — CHANGES_REQUESTED** [review]

**Pull request merged**